File: playwright-python/pages/admin_page.py
from playwright.sync_api import Page, expect


class AdminPage:
    def __init__(self, page: Page, base_url: str):
        self.page = page
        self.base_url = base_url.rstrip("/")
        self.username = page.get_by_label("Username")
        self.password = page.get_by_label("Password")
        self.login_button = page.get_by_role("button", name="Log in")

    def open(self) -> None:
        self.page.goto(f"{self.base_url}/admin/")

    def login(self, username: str, password: str) -> None:
        self.username.fill(username)
        self.password.fill(password)
        self.login_button.click()

    def expect_login_page(self) -> None:
        expect(self.username).to_be_visible()
        expect(self.password).to_be_visible()
        expect(self.login_button).to_be_visible()

    def expect_admin_index(self) -> None:
        expect(
            self.page.get_by_role("heading", name="Site administration")
        ).to_be_visible()

    def expect_products_and_orders(self) -> None:
        expect(self.page.get_by_role("link", name="Products", exact=True)).to_be_visible()
        expect(self.page.get_by_role("link", name="Orders", exact=True)).to_be_visible()

    def open_new_product_form(self) -> None:
        self.page.goto(f"{self.base_url}/admin/shop/product/add/")

    def fill_product_fields(self, product: dict[str, object]) -> None:
        self.page.get_by_label("Name:").fill(str(product["name"]))
        self.page.get_by_label("Description:").fill(str(product["description"]))
        self.page.get_by_label("Price:").fill(str(product["price"]))
        self.page.get_by_label("In stock:").fill(str(product["in_stock"]))

    def upload_product_image(self, image_path: str) -> None:
        self.page.get_by_label("Image:").set_input_files(image_path)

    def save_product(self) -> None:
        self.page.get_by_role("button", name="Save", exact=True).click()

    def open_product_change_form(self, product_id: int) -> None:
        self.page.goto(f"{self.base_url}/admin/shop/product/{product_id}/change/")

    def change_product_stock(self, stock: int) -> None:
        stock_field = self.page.get_by_label("In stock:")
        stock_field.fill(str(stock))
        self.save_product()

    def expect_product_listed(self, product_name: str) -> None:
        product_link = self.page.locator("#result_list").get_by_role(
            "link",
            name=product_name,
            exact=True,
        )
        expect(product_link).to_be_visible()

    def open_order_change_form(self, order_id: int) -> None:
        self.page.goto(f"{self.base_url}/admin/shop/order/{order_id}/change/")

    def expect_order_details(self, order: object) -> None:
        expect(self.page.get_by_label("Full name:")).to_have_value(order.full_name)
        expect(self.page.get_by_label("Email:")).to_have_value(order.email)
        expect(self.page.get_by_label("Phone:")).to_have_value(order.phone)
        expect(self.page.get_by_label("Address:")).to_have_value(order.address)
        expect(self.page.get_by_label("Postal code:")).to_have_value(order.postal_code)
        expect(self.page.get_by_label("City:")).to_have_value(order.city)
        expect(self.page.get_by_label("Country:")).to_have_value(order.country)

        content = self.page.locator("#content-main")
        expect(content).to_contain_text(str(order.total_amount))
        expect(content.locator(".field-created_at .readonly")).not_to_be_empty()

    def expect_order_items_read_only(self, items: list[object]) -> None:
        inline_group = self.page.locator("#items-group")
        for item in items:
            expect(inline_group).to_contain_text(item.product_name)
            expect(inline_group).to_contain_text(str(item.unit_price))
            expect(inline_group).to_contain_text(str(item.quantity))
            expect(inline_group).to_contain_text(str(item.line_total))

        for field_name in ("product_name", "unit_price", "quantity", "line_total"):
            expect(
                inline_group.locator(f'input[name$="-{field_name}"]')
            ).to_have_count(0)

File: playwright-typescript/src/pages/admin-page.ts
import { expect, type Locator, type Page } from "@playwright/test";
import type { Credentials, OrderItemRecord, OrderRecord } from "../support/models.js";
import { settings } from "../support/settings.js";

export class AdminPage {
  private readonly username: Locator;
  private readonly password: Locator;
  private readonly loginButton: Locator;

  constructor(private readonly page: Page) {
    this.username = page.getByLabel("Username");
    this.password = page.getByLabel("Password");
    this.loginButton = page.getByRole("button", { name: "Log in" });
  }

  open(): Promise<unknown> { return this.page.goto(`${settings.baseUrl}/admin/`); }

  async login(credentials: Credentials): Promise<void> {
    await this.username.fill(credentials.username);
    await this.password.fill(credentials.password);
    await this.loginButton.click();
  }

  async expectLoginPage(): Promise<void> {
    await expect(this.username).toBeVisible();
    await expect(this.password).toBeVisible();
    await expect(this.loginButton).toBeVisible();
  }

  expectIndex(): Promise<void> { return expect(this.page.getByRole("heading", { name: "Site administration" })).toBeVisible(); }

  async expectProductsAndOrders(): Promise<void> {
    await expect(this.page.getByRole("link", { name: "Products", exact: true })).toBeVisible();
    await expect(this.page.getByRole("link", { name: "Orders", exact: true })).toBeVisible();
  }

  openNewProduct(): Promise<unknown> { return this.page.goto(`${settings.baseUrl}/admin/shop/product/add/`); }

  async fillProduct(name: string, description: string, price: number, stock: number): Promise<void> {
    await this.page.getByLabel("Name:").fill(name);
    await this.page.getByLabel("Description:").fill(description);
    await this.page.getByLabel("Price:").fill(price.toFixed(2));
    await this.page.getByLabel("In stock:").fill(String(stock));
  }

  uploadImage(path: string): Promise<void> { return this.page.getByLabel("Image:").setInputFiles(path); }
  save(): Promise<void> { return this.page.getByRole("button", { name: "Save", exact: true }).click(); }
  openProduct(id: number): Promise<unknown> { return this.page.goto(`${settings.baseUrl}/admin/shop/product/${id}/change/`); }

  async changeStock(stock: number): Promise<void> {
    await this.page.getByLabel("In stock:").fill(String(stock));
    await this.save();
  }

  expectProductListed(name: string): Promise<void> {
    return expect(this.page.locator("#result_list").getByRole("link", { name, exact: true })).toBeVisible();
  }

  openOrder(id: number): Promise<unknown> { return this.page.goto(`${settings.baseUrl}/admin/shop/order/${id}/change/`); }

  async expectOrder(order: OrderRecord): Promise<void> {
    await expect(this.page.getByLabel("Full name:")).toHaveValue(order.full_name);
    await expect(this.page.getByLabel("Email:")).toHaveValue(order.email);
    await expect(this.page.getByLabel("Phone:")).toHaveValue(order.phone);
    await expect(this.page.getByLabel("Address:")).toHaveValue(order.address);
    await expect(this.page.getByLabel("Postal code:")).toHaveValue(order.postal_code);
    await expect(this.page.getByLabel("City:")).toHaveValue(order.city);
    await expect(this.page.getByLabel("Country:")).toHaveValue(order.country);
    await expect(this.page.locator("#content-main")).toContainText(String(order.total_amount));
    await expect(this.page.locator(".field-created_at .readonly")).not.toBeEmpty();
  }

  async expectItemsReadOnly(items: OrderItemRecord[]): Promise<void> {
    const group = this.page.locator("#items-group");
    for (const item of items) {
      await expect(group).toContainText(item.product_name);
      await expect(group).toContainText(String(item.unit_price));
      await expect(group).toContainText(String(item.quantity));
      await expect(group).toContainText(String(item.line_total));
    }
    for (const field of ["product_name", "unit_price", "quantity", "line_total"]) {
      await expect(group.locator(`input[name$="-${field}"]`)).toHaveCount(0);
    }
  }
}
